src/App.tsx

```diff
@@ -1,24 +1,284 @@
+    event.preventDefault();
+
+    if (!newTodoTitle.trim()) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    setErrorMessage('');
+    setIsLoadingAddTodo(true);
+
+    const newTodo = {
+      userId: USER_ID,
+      title: newTodoTitle.trim(),
+      completed: false,
+    };
+
+    setTempTodo({
+      ...newTodo,
+      id: 0,
+    });
+
+    addNewTodo(newTodo)
+      .then(response => {
+        setNewTodoTitle('');
+
+        return response;
+      })
+      .then(currentTodo => {
+        setTodos((prevState) => [...prevState, currentTodo]);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+      })
+      .finally(() => {
+        setIsLoadingAddTodo(false);
+        setTempTodo(null);
+      });
+  };
+
+  useEffect(() => {
+    setIsLoadingTodos(true);
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      })
+      .finally(() => {
+        setIsLoadingTodos(false);
+      });
+  }, []);
+
+  const handelDeleteTodoById = (todoId: number) => {
+    setTodosIdToDelete(prevState => [...prevState, todoId]);
+
+    return deleteTodoById(todoId)
+      .then(() => {
+        setTodos(prevState => prevState.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+        throw new Error();
+      })
+      .finally(() => {
+        setTodosIdToDelete(prevState => prevState.filter(id => id !== todoId));
+      });
+  };
+
+  const handleChangeCompletedStatus = (
+    todoId: number,
+    isCompleted: boolean,
+  ) => {
+    setTodosIdToUpdate(prevState => [...prevState, todoId]);
+
+    return updateTodoById(todoId, { completed: !isCompleted })
+      .then((currentTodo: Todo) => {
+        setTodos(prevState => getUpdatedTodos(
+          prevState,
+          currentTodo,
+          'completed',
+        ));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+        throw new Error();
+      })
+      .finally(() => {
+        setTodosIdToUpdate(prevState => prevState
+          .filter(id => id !== todoId));
+      });
+  };
+
+  const handleChangeTodoTitle = (
+    todoId: number,
+    newTitle: string,
+  ) => {
+    setTodosIdToUpdate(prevState => [...prevState, todoId]);
+
+    return updateTodoById(todoId, { title: newTitle })
+      .then((currentTodo) => {
+        setTodos(prevState => getUpdatedTodos(
+          prevState,
+          currentTodo,
+          'title',
+        ));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+        throw new Error();
+      })
+      .finally(() => {
+        setTodosIdToUpdate(prevState => prevState
+          .filter(id => id !== todoId));
+      });
+  };
+
+  const handelDeleteCompletedTodos = async () => {
+    const completedTodosId = todos
+      .filter(({ completed }) => completed)
+      .map(({ id }) => id);
```

Comment:
The same can be done through the `reduce` method and walk through the array once instead of twice.